design/design-doc-composition-functions.md

```diff
@@ -99,74 +99,78 @@ experiences designing Kubernetes, its precursor Borg, and many generations of
-and documentation. Either way, Crossplane has no religion - if you prefer not to
-“write YAML”, pick another tool and vice versa.
+* For advanced cases, use your DSL or GPL of choice.
+
+Note that the Composition logic (whether P&T, DSL, or GPL) is always "behind the
+API line" (behind an XR). This means the distinction is only important to the
+people _authoring_ the Compositions, never to the people consuming them.
+
+In the time since P&T Composition became available we've seen that:
+
+* Folks want to use Composition for more complex cases than we anticipated. Many
+  XRs fan out into tens of composed resources, sometimes nested.
+* Many Compositions call for a high level of expressiveness - conditionals,
+  iteration, merging data from multiple fields, etc.
+* The lack of a more expressive alternative to P&T Composition _has_ set us down
+  the path of organically growing a new DSL. (e.g. [#1972], [#2352], [#4051],
+  [#3917], [#3919], [#3989], [#3498], [#3458], [#3316], [#4036], [#4065],
+  [#4026])
+* Organically growing a new DSL is not only undesirable, but _slow_. Because
+  each addition to P&T Composition changes Crossplane's core API we must be
+  careful about what we accept. Changes take a long time to reach consensus.
+  They're coupled to Crossplane's release cycle, which means changes can take a
+  long time to become available, and aren't backported to older versions.
+* How folks would ideally configure Composition varies. Some folks prefer a
+  particular configuration DSL, others look for webhooks written in a GPL, etc.
+
+To address these issues, we added support for 'Composition Functions' in
+Crossplane v1.11. Functions are an alpha feature and are off by default. This
+document is an iteration on (and supersedes) the [previous design
+document][alpha-design] based on what we've learned during since the feature was
+launched.
 
 ## Goals
 
 The proposal put forward by this document should:
 
-* Let folks use their composition tool and/or programming language of choice.
 * Support 'advanced' composition logic such as loops and conditionals.
-* Balance safety (e.g. sandboxing) with speed and simplicity.
-* Be possible to introduce behind a feature flag that is off by default.
-
-While not an explicit goal, it would also be ideal if the solution put forth by
-this document could serve as a test bed for new features in the contemporary
-'resources array' based form of Composition.
-
-The user experience around authoring and maintaining Composition Functions is
-out of scope for this proposal, which focuses only on adding foundational
-support for the feature to Crossplane. 
+* Let folks specify composition logic in their DSL or GPL of choice.
+* Make it easy to extend Crossplane with new ways to 'do composition'.
+* Decouple adding new ways to 'do composition' from the core release cycle.
+* Be possible to keep behind a feature flag until it is generally available.
```

Comment:
Should we have some details on the motivation for the updated design? What was wrong with the previous one, and why do we want to change it? 
And/or a list of lessons learned from the alpha.

Reply:
We should. I struggled to enunciate them so I just skipped it. I think I have a better framing now - will update.